send mavlink Mission item reached

magicrub opened the pull request:
Send mavlink msg mission_item_reached when a mission item has completed.

fixes https://github.com/diydrones/ardupilot/issues/2169

WickedShell commented:
Looks good to me. (Haven't tried it out, just read through it).

magicrub commented:
I have not tested this, it's a copy/paste from @rmackay9 solution for copter. Seems simple enough though.

tridge commented:
this needs a check that there is enough buffer space on the channel before sending. If we don't check for available buffer space and there isn't enough then we end up losing both this message and potentially the next message (as the GCS tries to resync the stream). If there isn't enough buffer space it is better to just not send the message.
I also think we could move the looping logic into libraries/GCS_MAVLink, similar to send_statustext_all()

magicrub commented:
redesigning to use message try queue. Will resubmit rover, copter and plane under a new single PR

magicrub commented:
Ok, I've completely re-wrote this so it uses the try_send que method. it also has a static variable to capture the correct index at time of sending initial index. If it was to try to resend later the index would have changed so this was required in some form

magicrub commented:
Changes:
- rebased on master
- moved more code into GCS_Common
- replaced static index with class member
- unified across all 4 platforms

TODO: As soon as RAM is not a critical as it is now for APM 2.x platforms, I propose we change the "try" stuff to use a circular buffer which is a class that includes a param. That will ensure all msgs get sent as you wish them to get sent and none will get lost when additional msgs come in of the same type

rmackay9 commented:
I'm not sure if it's critical or not but this new method will not send a message when a do-command is completed like the current method implemented in master for copter.

magicrub commented:
It would need to report reached of all mission item types to be correct.
I'll take a look.
On Jul 24, 2015 7:10 PM, "Randy Mackay" <notifications@github.com> wrote:

> I'm not sure if it's critical or not but this new method will not send a
> message when a do-command is completed like the current method implemented
> in master for copter.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/diydrones/ardupilot/pull/2573#issuecomment-124788780>.
>

tridge commented on APMrover2/commands_logic.cpp:
```diff
@@ -122,6 +122,25 @@ void Rover::exit_mission()
+        return cmd_complete;
+    }
+
+    // exit immediately if not in AUTO mode
```
copter and plane return false here. Maybe it doesn't matter?

magicrub replied:
agreed, all should return false. However, as Randy as noted this callback can't even get called when not in auto so the whole control_mode == AUTO check, and the else return false are useless code. If we want to remove any of that lets do it in a separate PR

tridge commented:
I think there is a problem where if there are deferred messages at the time that gcs_send_message(MSG_MISSION_ITEM_REACHED) is called, and the serial port is full, then it will end up sending a MISSION_ITEM_REACHED with the next wp number, not the one that has just been reached.
I think the fix is to set _mission_item_reached_index at the time gcs_send_message() is called, and not rely on it being set by try_send_message(). The try_send_message() function is not guaranteed to be called immediately, and if it isn't then the current wp in the mission object will have moved on, and we'll end up sending a message for the wrong wp

magicrub commented:
@randy and @tridge your suggestions have been implemented. the index is stored at the callback time. Also rover is changed to "false". I rebase interactively the plane changes but left rover and copter changes ad additions so its easier to see what I changed

magicrub commented:
@rmackay9 I tested this with some DO_SET_SERVO commands and they reported as expected but the DO_JUMP doesn't and thats because there's nothings actually executing to trigger a reached so I'm ok with that.

magicrub commented:
BUMP - rebased on master

tridge commented:
applied with minor changes as discussed
thanks!
